src/cpp/session/prefs/UserStateLayer.cpp

```diff
@@ -33,18 +33,34 @@ UserStateLayer::UserStateLayer():
 
 core::Error UserStateLayer::readPrefs()
 {
-   prefsFile_ = core::system::xdg::userDataDir().completePath(
+   FilePath schemaFile = options().rResourcesPath().completePath("schema").completePath(kUserStateSchemaFile);
+
+   // desktop and server versions of RStudio use separate state files so that mixing desktop and server
+   // on the same machine is possible w/o side effects like sharing a source database
+   stateFile_ = core::system::xdg::userDataDir().completePath(
          options().programMode() == kSessionProgramModeDesktop ? 
             kUserStateFileDesktop : 
             kUserStateFileServer);
 
-   return loadPrefsFromFile(prefsFile_,
-      options().rResourcesPath().completePath("schema").completePath(kUserStateSchemaFile));
+   if (!stateFile_.exists())
+   {
+      // if there's no state file yet, check for a state file left by an older version of RStudio 1.3
+      FilePath oldStateFile = core::system::xdg::userDataDir().completePath("rstudio-state.json");
+      if (oldStateFile.exists())
+      {
+          // found an old file; attempt to migrate it
+          Error error = oldStateFile.move(stateFile_);
```

Comment:
So in pathological case where (for example) user was last using RStudio Server 1.3 (prior to the separation of desktop vs server state files), then installs updated RStudio Desktop and Server 1.3 and runs desktop, it will take over that previous state, and next time they run Server that state is gone? Maybe nothing can be done about this. What if you, on first time encountering this scenario where the new state filename is missing, copy the legacy state file over both to desktop and server then delete the old one? Maybe that would be worse, I'm not sure of the implications or if would be worth the complexity.

Reply:
Your analysis is 100% correct; if you are running both desktop and server then only one of them is going to get your old state. That's unfortunate but one of them has to win since we can't use the same state for both of them:

> What if you, on first time encountering this scenario where the new state filename is missing, copy the legacy state file over both to desktop and server then delete the old one?

This would be worse. Copying it to both locations means that desktop and server will *both* use the old context ID, which means they will share a source database. (see https://github.com/rstudio/rstudio/pull/6669).